### PR TITLE
Async-from-sync test the iterator closes when .throw() is undefined

### LINKED ISSUE #3387: Tests for  Async-from-Sync Iterator changes
https://github.com/tc39/ecma262/pull/2600 reached consensus at the last plenary. It updates  Async-from-Sync Iterator objects to close the wrapped iterator in cases where the async wrapper is considered to be done, but the sync iterator isn't.

I included some examples in the presentation of where this behavior shows up, and there were a few relevant cases in the linked issue. I plan to adapt those into test262 cases, but I may not have time immediately, so creating this issue to keep track of the tests needed.

### PR DESCRIPTION
**Pull request by ioannad**
This adds tests for the normative changes in https://github.com/tc39/ecma262/pull/2600 for the new steps in ` %AsyncFromSyncIteratorPrototype%.throw`, also modifying two existing tests.

These tests have been tested with this PR of engine262, which implements the normative changes: https://github.com/engine262/engine262/pull/230

Partly fixes #3387

### CODE REVIEW

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js
Comment (nicolo-ribaudo) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js:57]: The spec doesn't define error messages, so we cannot test this.

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js:56]:
```suggestion
  assert.sameValue(err instanceof TypeError, true);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js:69]:
```suggestion
      assert.sameValue(err instanceof TypeError, true);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js:74]:
```suggestion
      iter.next().then(function (result) {
        assert.sameValue(result.done, true, 'the iterator is completed');
        assert.sameValue(result.value, undefined, 'value is undefined');
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js:79]:
```suggestion
  ).then($DONE, $DONE);

}).then($DONE, $DONE);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-poisoned-return.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-poisoned-return.js:77]:
```suggestion
      iter.next().then(function (result) {
        assert.sameValue(result.done, true, 'the iterator is completed');
        assert.sameValue(result.value, undefined, 'value is undefined');
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-poisoned-return.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-poisoned-return.js:82]:
```suggestion
  ).then($DONE, $DONE);

}).then($DONE, $DONE);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js:70]:
```suggestion
iter.next().then(function (result) {
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js:77]:
```suggestion
      assert.sameValue(err instanceof TypeError, true);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js:82]:
```suggestion
      iter.next().then(function (result) {
        assert.sameValue(result.done, true, 'the iterator is completed');
        assert.sameValue(result.value, undefined, 'value is undefined');
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-not-object.js:87]:
```suggestion
  ).then($DONE, $DONE);

}).then($DONE, $DONE);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js:71]:
```suggestion
      assert.sameValue(err instanceof TypeError, true);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js:76]:
```suggestion
      iter.next().then(function (result) {
        assert.sameValue(result.done, true, 'the iterator is completed');
        assert.sameValue(result.value, undefined, 'value is undefined');
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-return-object.js:81]:
```suggestion
  ).then($DONE, $DONE);

}).then($DONE, $DONE);
```

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined.js
Comment (ljharb) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined.js:46]:
```suggestion
      assert.sameValue(err instanceof TypeError, true);
```

**Review by ljharb — DISMISSED** [review]

**Comment by ioannad** [comment]
Thank you @nicolo-ribaudo  and @ljharb for the review and suggestions!

**Comment by bakkot** [comment]
@Ms2ger Re: the label: the PR to the specification has had approval for ~2 years and is only waiting on these tests to land (i.e. it's conceptually "stage 3"). If you're just waiting on the specification, and these are otherwise ready to land, they should be landed, I think.

We can land the spec PR first if necessary, but I don't want to do that until the tests are actually ready to land.

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js
Comment (ptomato) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js:60]: Optional, but if you like - it might be a good opportunity to refactor these tests to use the [Async Helpers](https://github.com/tc39/test262/blob/main/rfcs/async-helpers.md). However, it's fine with me to land this PR without doing that.

As an example, I think this test would become something like
```js
asyncTest(async function () {
  await assert.throwsAsync(TypeError, () => asyncIterator.next(), "Promise should be rejected");
  const result = await asyncIterator.next();
  assert(result.done, "the iterator is completed");
  assert.sameValue(result.value, undefined, "value is undefined");
});
```
...which seems to me easier to follow.

**Review by ptomato — APPROVED** [review]
I did not check these line by line against the normative change. So it would be nice-to-have if someone more familiar with that change would take a look. Even without that, I think the use of engine262 gives sufficient confidence and I would be fine with landing it as is.

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js
Comment (ptomato) [inline test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js:79]: I didn't notice this the first time but I noticed @nicolo-ribaudo's comment on #3977 — applies here and in the other files too. (Using the async helpers would save you from having to call `$DONE` at all.)

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js
Comment (ioannad) [reply test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-null.js]: The latest commit refactors the tests to use Async Helpers as suggested.

File: test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js
Comment (ioannad) [reply test/built-ins/AsyncFromSyncIteratorPrototype/throw/throw-undefined-get-return-undefined.js]: The latest commit refactors the tests to use Async Helpers as suggested.

**Review by ptomato — APPROVED** [review]
:+1:

**Comment by ptomato** [comment]
All comments have been addressed and no new ones have shown up since last week, let's let this train leave the station! :smile:

**Pull request merged**
